src/QueueReceiver.Core/Services/PersonProjectService.cs

```diff
@@ -64,14 +74,49 @@ public async Task GiveProjectAccessToPlant(long personId, string plantId)
                 var restrictionRole = await _restrictionRoleRepository.FindRestrictionRole(DefaultRestrictionRole, plantId);
                 await _personRestrictionRoleRepository.AddIfNotExistAsync(plantId, restrictionRole, personId);
 
+                projects.ForEach(p =>
+                {
+                    _personProjectHistoryService.LogAddAccess(personId, personProjectHistory, p.ProjectId);
+                    
+                    if (p.IsMainProject)
+                    {
+                        _personProjectHistoryService.LogDefaultUserGroup(personId, personProjectHistory, p.ProjectId);
+                    }
+                });
+            }
+
+            if (unvoided)
+            { 
+                projects.ForEach(p => _personProjectHistoryService.LogUnvoidProjects(personId, personProjectHistory, p.ProjectId));
+            }
+
+            if (unvoided || updated)
+            {
+                await _personProjectHistoryRepository.AddAsync(personProjectHistory);
                 await _personProjectRepository.SaveChangesAsync();
             }
         }
 
         public async Task RemoveAccessToPlant(long personId, string plantId)
         {
+            var personProjectHistory = CreatePersonProjectHistory(personId);
+            var projects = await _projectRepository.GetParentProjectsByPlant(plantId);
             _personProjectRepository.VoidPersonProjects(plantId, personId);
+
+            projects.ForEach(p => _personProjectHistoryService.LogVoidProjects(personId, personProjectHistory, p.ProjectId));
+            
             await _personProjectRepository.SaveChangesAsync();
         }
+
+        private PersonProjectHistory CreatePersonProjectHistory(long personId)
```

Comment:
Move this to PersonProjectHistoryService. Or possibly just make this the constructor with these parameters and scrap the method.